### PR TITLE
[CYTHON] Cython module for NDArray

### PR DESCRIPTION
**Pull request by tqchen**
@piiswrong please review

### CODE REVIEW

**Comment by tqchen** [comment]
@jermainewang I changed the setuptools back to distutil, because although the setuptool enables correct installation, the build --inplace did not copy the .so file of corresponding module correctly back to the python/mxnet/_cy2 folder, so the PYTHONPATH way does not work

File: include/mxnet/c_api.h [hunk 1]
```diff
@@ -486,13 +486,21 @@ MXNET_DLL int MXImperativeInvoke(AtomicSymbolCreator creator,
...
 //--------------------------------------------
 // Part 3: symbolic configuration generation
 //--------------------------------------------
 /*!
+ * \brief list all the available operator names, include entries.
+ * \param out_size the size of returned array
+ * \param out_array the output operator name array.
+ * \return 0 when success, -1 when failure happens
+ */
+MXNET_DLL int MXListAllOpNames(mx_uint *out_size,
```
Comment (piiswrong) [inline include/mxnet/c_api.h]: Do we need this? Can't it be merged with symbolic list?

File: python/mxnet/cython/ndarray.pyx [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+                    for item in v:
+                        if not isinstance(item, NDArrayBase):
+                            raise ValueError("out need to be of type NDArray")
+                        output_vars.push_back((<NDArrayBase>v).chandle)
+            else:
+                sparam_vals.push_back(c_str(str(v)))
+                sparam_keys.push_back(c_str(k))
+
+        num_output = output_vars.size()
+        if output_vars.size() == 0:
```
Comment (piiswrong) [inline python/mxnet/cython/ndarray.pyx]: If out is empty it should be alloced by the backend

Comment (tqchen) [inline python/mxnet/cython/ndarray.pyx]: fixed

**Pull request merged**